R/stacking.R

```diff
@@ -0,0 +1,1626 @@
+            next
+        }
+
+        # Check for overlap with previous manual stacking inputs
+        if (length(manual.stacking.groups) > 0)
+        {
+            group.ind.without.na <- removeNA(group.ind)
+            overlap.ind <- which(vapply(manual.stacking.groups, function(manual.group) {
+                any(group.ind.without.na %in% manual.group)
+            }, logical(1)))
+            if (length(overlap.ind) > 0)
+            {
+                warning("The manual stacking input '", input.text,
+                        "' has been ignored as it contains variable(s) that ",
+                        "overlap with another manual stacking input '",
+                        manual.stacking.groups.text[overlap.ind[1]], "'.")
+                next
+            }
+        }
+
+        manual.stacking.groups <- c(manual.stacking.groups, list(group.ind))
+        manual.stacking.groups.text <- c(manual.stacking.groups.text,
+                                         input.text)
+    }
+    if (length(manual.stacking.groups) == 0)
+        return(stacking.groups)
+
+    if (is.null(stacking.groups))
+        stacking.groups <- matrix(nrow = 0, ncol = 0)
+
+    # Remove rows in stacking.groups that contain overlap with
+    # manual.stacking.groups
+    for (manual.group in manual.stacking.groups)
+    {
+        manual.group.without.na <- removeNA(manual.group)
+        is.overlapping <- apply(stacking.groups, 1, function(group) {
+            any(manual.group.without.na %in% group)
+        })
+        stacking.groups <- stacking.groups[!is.overlapping, , drop = FALSE]
+    }
+
+    # Increase number of columns in stacking.groups if necessary to fit
+    # manual.stacking.groups
+    n.rows <- nrow(stacking.groups)
+    n.cols <- ncol(stacking.groups)
+    max.stacking <- max(max(vapply(manual.stacking.groups, length, integer(1))),
+                        n.cols)
+    new.stacking.groups <- matrix(nrow = n.rows, ncol = max.stacking)
+    new.stacking.groups[seq_len(n.rows), seq_len(n.cols)] <- stacking.groups
+    stacking.groups <- new.stacking.groups
+
+    # Append manual.stacking.groups to stacking.groups
+    for (manual.group in manual.stacking.groups)
+    {
+        new.group <- rep(NA_integer_, max.stacking)
+        new.group[seq_along(manual.group)] <- manual.group
+        stacking.groups <- rbind(stacking.groups, new.group)
+    }
+
+    n.manual.stacking <- length(manual.stacking.groups)
+    is.manually.stacked <- c(rep(FALSE, nrow(stacking.groups) - n.manual.stacking),
+                             rep(TRUE, n.manual.stacking))
+    attr(stacking.groups, "is.manually.stacked") <- is.manually.stacked
+
+    stacking.groups
+}
+
+# Parse user-input manual stacking strings for stacking specified by
+# observation, i.e., each string in manual.stacking describes the variables in
+# an observation to be used in stackings of variables.
+stackingSpecifiedByObservation <- function(stacking.groups, manual.stacking,
```

Comment:
I don't get the difference between this and stackingSpecifiedByVariable.  Unit tests and examples would have helped.

Reply:
Added unit tests